pkg/controller/controller_utils.go

```diff
@@ -542,6 +544,39 @@ func GetPodFromTemplate(template *v1.PodTemplateSpec, parentObject runtime.Objec
 	return pod, nil
 }
 
+func (r RealPodControl) validateNodeConflict(pod *v1.Pod) error {
+	if len(pod.Spec.NodeName) == 0 {
+		return nil
+	}
+
+	node, err := r.KubeClient.Core().Nodes().Get(pod.Spec.NodeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	nodeInfo := schedulercache.NewNodeInfo()
+	err = nodeInfo.SetNode(node)
+	if err != nil {
+		return err
+	}
+
+	fit, reasons, err := predicates.PodSelectorMatches(pod, nil, nodeInfo)
```

Comment:
As @xiangpengzhao alluded to, the problem is that anyone can set NodeName -- end users, controllers, scheduler, etc. There are pros and cons of this (mostly cons IMO) but it is too late to change this behavior because of backward compatibility concerns (at least until we do a full rev of the API, e.g. go to v2 API).

Because any person or component can set NodeName on a pod, it's necessary for Kubelet to run the scheduler feasibility checks as part of the admission process to "protect itself" from various bad things (e.g. overcommitting request on a node).

I talked with @bsalamat about this and we both think that the best/only thing we can do here is to have the controller notice that the pods it is creating keep getting rejected by kubelet with a "reason" that indicates kubelet admission failed, and rate limit creation of subsequent pods (until they start succeeding, at which time it would stop the rate limiting).